qcodes/tests/dataset/test_data_set_cache.py

```diff
@@ -0,0 +1,297 @@
+                    # TODO the expansion both for cache and get_parameter_data is buggy iff all types are array
+                    assert data[subparam][DAC.ch1.full_name].shape == expected_shape
+                for setpoint_name in param.setpoint_full_names[j]:
+                    assert data[subparam][setpoint_name].shape == expected_shape
+            _assert_parameter_data_is_identical(dataset.get_parameter_data(),
+                                                data)
+    _assert_parameter_data_is_identical(dataset.get_parameter_data(),
+                                        dataset.cache.data())
+
+
+@pytest.mark.parametrize("bg_writing", [True, False])
+@pytest.mark.parametrize("storage_type", ['array', None])
+@pytest.mark.parametrize("outer_param_type", ['numeric', 'text'])
+@settings(deadline=None, max_examples=10)
+@given(n_points=hst.integers(min_value=1, max_value=21))
+def test_cache_complex_array_param_in_1d(experiment, DAC, channel_array_instrument, n_points, bg_writing, storage_type, outer_param_type):
+    param = channel_array_instrument.A.dummy_complex_array_parameter
+    meas = Measurement()
+    if outer_param_type == 'numeric':
+        outer_param = DAC.ch1
+        outer_setpoints = np.linspace(-1, 1, n_points)
+        outer_storage_type = storage_type
+    else:
+        outer_param = channel_array_instrument.A.dummy_text
+        outer_setpoints = ['A', 'B', 'C', 'D']
+        outer_storage_type = 'text'
+    meas.register_parameter(outer_param, paramtype=outer_storage_type)
+    meas.register_parameter(param, setpoints=(outer_param,), paramtype=storage_type)
+    array_used = _array_param_used_in_tree(meas)
+    with meas.run(write_in_background=bg_writing) as datasaver:
+        dataset = datasaver.dataset
+        for i, v1 in enumerate(outer_setpoints):
+            datasaver.add_result((outer_param, v1),
+                                 (param, param.get()))
+            datasaver.flush_data_to_database(block=True)
+            data = dataset.cache.data()
+            n_rows_written = i+1
+
+            if array_used:
+                expected_shape = (n_rows_written,) + param.shape
+            else:
+                expected_shape = n_rows_written * np.prod(param.shape)
+            assert data[param.full_name][param.full_name].shape == expected_shape
+            if storage_type != 'array':
+                # TODO the expansion both for cache and get_parameter_data is buggy iff all types are array
+                assert data[param.full_name][outer_param.full_name].shape == expected_shape
+            for setpoint_name in param.setpoint_full_names:
+                assert data[param.full_name][setpoint_name].shape == expected_shape
+            _assert_parameter_data_is_identical(dataset.get_parameter_data(),
+                                                data)
+    _assert_parameter_data_is_identical(dataset.get_parameter_data(),
+                                        dataset.cache.data())
+
+
```

Comment:
i couldn't see if there is a test where dependent parameters z1 and z2 are both dependent on x and y but, say, z1 is written for 10 values of x, and z2 is written for 20 values of x? did i miss it or does it not exist? if it doesn't, it seems like a worthwhile test to ensure that tracking rows per param tree works well.

Reply:
Its there indirectly due to the multiparameters etc but lets just add it explicit too